### PR TITLE
Fixed support of 0-sized arrays for 'linalg.qr'

### PR DESCRIPTION
**Pull request by IvanYashchuk**
With reference to #2371 

Moreover, cuSOLVER does not calculate Q correctly if R is zero-sized. This case is handled separately.

### CODE REVIEW

**Comment by niboshi** [comment]
Thanks! Jenkins, test this please

**Comment by niboshi** [comment]
It seems Jenkins test failures are caused by this fix.
Could you check them out?

**Comment by IvanYashchuk** [comment]
Jenkins environment seems to have NumPy version older than 1.16.

**Comment by niboshi** [comment]
@IvanYashchuk Sorry fo delay. Yes I think so. Let's skip those in `numpy<1.16`.

File: tests/cupy_tests/linalg_tests/test_decomposition.py [hunk 1]
```diff
@@ -89,12 +91,19 @@ def check_mode(self, array, mode, dtype):
...
-        self.check_mode(numpy.random.randn(3, 0), mode=self.mode)
         self.check_mode(numpy.random.randn(2, 4), mode=self.mode)
         self.check_mode(numpy.random.randn(3, 3), mode=self.mode)
         self.check_mode(numpy.random.randn(5, 4), mode=self.mode)
 
+    @testing.fix_random()
+    @condition.repeat(3, 10)
+    @unittest.skipUnless(
+        LooseVersion(numpy.__version__) > LooseVersion('1.16.0'),
+        'NumPy<1.16 does not work with empty arrays')
```
Comment (niboshi) [inline tests/cupy_tests/linalg_tests/test_decomposition.py]: There is a decorator for this: `cupy.testing.with_requires`.
Let's use it.

Comment (IvanYashchuk) [reply tests/cupy_tests/linalg_tests/test_decomposition.py]: Done.

**Comment by niboshi** [comment]
Sorry, I overlooked [the comment](https://github.com/cupy/cupy/pull/2372#discussion_r313864232).
Jenkins, test this please

**Comment by niboshi** [comment]
Could you check the test results?

File: tests/cupy_tests/linalg_tests/test_decomposition.py [hunk 2]
```diff
@@ -93,6 +93,13 @@ def test_mode(self):
         self.check_mode(numpy.random.randn(3, 3), mode=self.mode)
         self.check_mode(numpy.random.randn(5, 4), mode=self.mode)
 
+    @testing.fix_random()
+    @condition.repeat(3, 10)
```
Comment (toslunar) [inline tests/cupy_tests/linalg_tests/test_decomposition.py]: There's no randomness in the inputs.  Could you remove `fix_random` and `repeat`?  `numpy.empty((0, 3))` can be used below.

**Comment by toslunar** [comment]
The issue for `linalg.qr` is fixed by #2586.  @IvanYashchuk, thanks for your initial work.

**Pull request closed without merging**